web-service/helpers.py

```diff
@@ -14,140 +14,119 @@
-                device_data["id"] = url
-                device_data["url"] = siteInfo.url
-                if siteInfo.title is not None:
-                    device_data["title"] = siteInfo.title
-                if siteInfo.description is not None:
-                    device_data["description"] = siteInfo.description
-                if siteInfo.favicon_url is not None:
-                    device_data["icon"] = siteInfo.favicon_url
-                if siteInfo.jsonlds is not None:
-                    device_data["json-ld"] = json.loads(siteInfo.jsonlds)
-            else:
-                device_data["id"] = url
-                device_data["url"] = url
-
-            metadata_output.append(device_data)
-
-        return metadata_output
+    metadata_output = []
+
+    # Resolve the devices
+    for obj in objects:
+        key_id = None
+        url = None
+        force = False
+        valid = True
+        siteInfo = None
+        rssi = None
+        txpower = None
+
+        if "id" in obj:
+            key_id = obj["id"]
+        elif "url" in obj:
+            key_id = obj["url"]
+            url = obj["url"]
+            parsed_url = urlparse(url)
+            if parsed_url.scheme != 'http' and parsed_url.scheme != 'https':
+                valid = False
+
+        if "force" in obj:
+            force = True
+
+        # We need to go and fetch.  We probably want to asyncly fetch.
+
+        try:
+            rssi = float(obj["rssi"])
+            txpower = float(obj["txpower"])
+        except:
+            pass
+
+        if valid:
+            # Really if we don't have the data we should not return it.
+            siteInfo = models.SiteInformation.get_by_id(url)
+
+            if force or siteInfo is None:
+                # If we don't have the data or it is older than 5 minutes, fetch.
+                siteInfo = FetchAndStoreUrl(siteInfo, url)
+            if siteInfo is not None and siteInfo.updated_on < datetime.now() - timedelta(minutes=5):
+                # Updated time to make sure we don't request twice.
+                siteInfo.put()
+                # Add request to queue.
+                taskqueue.add(url='/refresh-url', params={'url': url})
+
+        device_data = {};
+        if siteInfo is not None:
+            device_data["id"] = url
+            device_data["url"] = siteInfo.url
+            if siteInfo.title is not None:
+                device_data["title"] = siteInfo.title
+            if siteInfo.description is not None:
+                device_data["description"] = siteInfo.description
+            if siteInfo.favicon_url is not None:
+                device_data["icon"] = siteInfo.favicon_url
+            if siteInfo.jsonlds is not None:
+                device_data["json-ld"] = json.loads(siteInfo.jsonlds)
+        else:
+            device_data["id"] = url
+            device_data["url"] = url
+        device_data["rssi"] = rssi
+        device_data["txpower"] = txpower
```

Comment:
I meant a rank value or a score.
Device will be sent in individual requests.
Therefore, at each request, it will sort one device individually.
And there will be no way for the client to sort them.